lib/template-parts.php

```diff
@@ -177,92 +170,103 @@ function filter_rest_wp_template_part_collection_params( $query_params ) {
-			$template_part_files = get_template_part_paths( get_stylesheet_directory() );
-			if ( is_child_theme() ) {
-				$template_part_files = array_merge( $template_part_files, get_template_part_paths( get_template_directory() ) );
-			}
-			// Build and save each template part.
-			foreach ( $template_part_files as $template_part_file ) {
-				$content = file_get_contents( $template_part_file );
-				// Infer slug from filepath.
-				$slug = substr(
-					$template_part_file,
-					// Starting position of slug.
-					strpos( $template_part_file, 'block-template-parts/' ) + 21,
-					// Subtract ending '.html'.
-					-5
-				);
-				// Wrap content with the template part block, parse, and create auto-draft.
-				$template_part_string = '<!-- wp:template-part {"slug":"' . $slug . '","theme":"' . wp_get_theme()->get( 'TextDomain' ) . '"} -->' . $content . '<!-- /wp:template-part -->';
-				$template_part_block  = parse_blocks( $template_part_string )[0];
-				create_auto_draft_for_template_part_block( $template_part_block );
-			}
-		};
-
 		$args['meta_query'] = $meta_query;
 	}
 
 	return $args;
 }
 add_filter( 'rest_wp_template_part_query', 'filter_rest_wp_template_part_query', 99, 2 );
+
+/**
+ * Creates a template part auto-draft if it doesn't exist yet.
+ *
+ * @access private
+ *
+ * @param string $slug    Template part slug.
+ * @param string $theme   Template part theme.
+ * @param string $content Template part content.
+ */
+function create_auto_draft_for_template_part( $slug, $theme, $content ) {
+	// We check if an auto-draft was already created,
+	// before running the REST API calls
+	// because the site editor needs an existing auto-draft
+	// for each theme template part to work properly.
+	$template_part_query = new WP_Query(
+		array(
+			'post_type'      => 'wp_template_part',
+			'post_status'    => array( 'publish', 'auto-draft' ),
+			'title'          => $slug,
+			'meta_key'       => 'theme',
+			'meta_value'     => $theme,
+			'posts_per_page' => 1,
+			'no_found_rows'  => true,
+		)
+	);
+	$template_part_post  = $template_part_query->have_posts() ? $template_part_query->next_post() : null;
+	if ( ! $template_part_post ) {
+		wp_insert_post(
+			array(
+				'post_content' => $content,
+				'post_title'   => $slug,
+				'post_status'  => 'auto-draft',
+				'post_type'    => 'wp_template_part',
+				'post_name'    => $slug,
+				'meta_input'   => array(
+					'theme' => $theme,
+				),
+			)
+		);
+	} else {
+		// Potentially we could decide to update the content if different.
+	}
+}
+
+/**
+ * Create the template parts auto-drafts for the current theme.
+ *
+ * @access private
+ */
+function synchronize_theme_template_parts() {
+	/**
+	 * Finds all nested template part file paths in a theme's directory.
+	 *
+	 * @param string $base_directory The theme's file path.
+	 * @return array $path_list A list of paths to all template part files.
+	 */
+	function get_template_part_paths( $base_directory ) {
```

Comment:
> Also could this reused with synchronize_theme_templates by adding the name of folder as param?

There's definitely some factorization to do here indeed.

The function inside a function is just a way to avoid exposing functions in the global scope.

Reply:
> The function inside a function is just a way to avoid exposing functions in the global scope.

PHP doesn't support this. https://3v4l.org/ZUp15

Reply:
Interesting :)

Reply:
> PHP doesn't support this.

Wow!  TIL...